### PR TITLE
fix: replace Webpacker with Shakapacker

### PR DESCRIPTION
**Pull request by adriangohjw**
### Summary

closes https://github.com/shakacode/react_on_rails/issues/1621 by replacing `Webpacker.config.shakapacker_precompile?` in configuration to `Shakapacker.config.shakapacker_precompile?`

### Pull Request checklist
_Remove this line after checking all the items here. If the item is not applicable to the PR, both check it out and wrap it by `~`._

- [x] Add/update test to cover these changes
- [ ] ~Update documentation~
- [x] Update CHANGELOG file 

### Other Information

_Remove this paragraph and mention any other important and relevant information such as benchmarks._

- - -
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1622)

## Summary by CodeRabbit

- **New Features**
	- Added dynamic handling for `shakapacker` versions in workflows.
	- Enhanced error handling related to `shakapacker`.

- **Refactor**
	- Renamed `WebpackerUtils` module to `PackerUtils` with updated methods and logic.

- **Chores**
	- Updated Ruby and Node versions in GitHub Actions workflows.
	- Simplified workflow matrices and caching strategies.

- **Documentation**
	- Updated contributing guidelines to reflect changes in the development environment setup.

### CODE REVIEW

**Comment by justin808** [comment]
@Judahmeek please review.

**Comment by justin808** [comment]
@adriangohjw can you please add a changelog entry.

**Review by Judahmeek — APPROVED** [review]
LGTM

**Comment by adriangohjw** [comment]
> @adriangohjw can you please add a changelog entry.

@justin808 oops my bad, saw that it has been addressed in https://github.com/shakacode/react_on_rails/pull/1622/commits/9ae6922153b8144c9d36553ae12434f23bdaf987 - thanks @Judahmeek !

**Comment by justin808** [comment]
I'll ship once CI passes!

**Comment by justin808** [comment]
@adriangohjw please check CI failures.

**Comment by Judahmeek** [comment]
@adriangohjw Shakapacker v6 & v8 are incompatible since v6 uses the `Webpacker` constant. Shakapacker v7 contains both `Shakapacker` & `Webpacker` constants & should be compatible with v8, so let's just drop support for v6.

**Comment by adriangohjw** [comment]
@Judahmeek 
- i'm not sure if i'm making the correct changes
- i'm also unsure if I should change `WebpackerUtils` to `ShakapackerUtils`

can you advise? thank you!

**Comment by Judahmeek** [comment]
> * i'm not sure if i'm making the correct changes
Your changes look promising. We've got to get CI to pass though.
> * i'm also unsure if I should change `WebpackerUtils` to `ShakapackerUtils`
Yes, please do.

**Review by Judahmeek — COMMENTED** [review]
Need CI to pass.

**Comment by adriangohjw** [comment]
@Judahmeek 
- CI good to go (Reference: https://github.com/adriangohjw/react_on_rails/pull/2) + Addressed comments from the automated code review
- given that we are dropping support for shakapacker v6, perhaps the new version should not be `14.0.2`?

**Comment by Judahmeek** [comment]
> * CI good to go (Reference: [Fix uninitialized constant webpacker adriangohjw/react_on_rails#2](https://github.com/adriangohjw/react_on_rails/pull/2)) + Addressed comments from the automated code review

I greatly appreciate your work on that. 👍 
> * given that we are dropping support for shakapacker v6, perhaps the new version should not be `14.0.2`?

Correct. Any time we change supported dependency version ranges, then the next release of ReactOnRails will be a  new major version since it is a breaking change.

**Comment by adriangohjw** [comment]
@Judahmeek any update on this? I think the workflow got stuck (might be due to the Github actions incident last week - https://www.githubstatus.com/incidents/nj47vccwm2zj)

**Comment by vaukalak** [comment]
@justin808 I've tested this branch, and it did work fine.

**Comment by timkrins** [comment]
Thanks for this - I gave it a try locally.

The file `lib/generators/react_on_rails/templates/base/base/config/shakapacker.yml` generates config that is no longer in-line with the Shackapacker v8 default `lib/install/config/shakapacker.yml` file.

After running `rails generate react_on_rails:install`, and running `./bin/dev`, Webpack (`webpack-cli` 5.1.4) does not compile, instead raising:
```
wp-client | [webpack-cli] Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
wp-client |  - options has an unknown property 'https'. These properties are valid:
wp-client |    object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, ipc?, liveReload?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? }
wp-client | Exited with code 2
```

This is because Webpack dropped the `https: true` key in favor of `server: https`.
The config in the Shackapacker `shakapacker.yml` were changed in https://github.com/shakacode/shakapacker/pull/277

**Comment by justin808** [comment]
> > This is because Webpack dropped the https: ... key in favor of server: https.
> 
> thanks @timkrins - will make the change
> 
> @justin808 should we copy-paste the config from https://github.com/shakacode/shakapacker/blob/main/lib/install/config/shakapacker.yml?

Yes, as it's the current one and the current version is the default.

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,14 @@ Please follow the recommendations outlined at [keepachangelog.com](http://keepac
 ### [Unreleased]
 Changes since the last non-beta release.
 
+### [14.0.2] - 2024-05-20
+
+#### Removed
+- Dropped Shakapacker v6 support [PR 1622](https://github.com/shakacode/react_on_rails/pull/1622) by [adriangohjw](https://github.com/adriangohjw)
```
Comment (justin808) [inline CHANGELOG.md]: This would be a breaking change and require going to v15.

could we still support shakapacker v6? 

@Judahmeek ?

Comment (vaukalak) [reply CHANGELOG.md]: @justin808 yes, creating a POC PR for that

**Comment by Judahmeek** [comment]
@adriangohjw Once again, thanks for your work.

Justin decided that he wanted to use an adapter module instead of having to make any breaking changes so I'm incorporating work from an old PR of mine.

File: Gemfile.development_dependencies [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-gem "shakapacker", "7.2.1"
+gem "shakapacker", "8.0.0"
```
Comment (justin808) [inline Gemfile.development_dependencies:3]: update to latest

File: lib/react_on_rails/packer_utils.rb [hunk 3]
```diff
@@ -1,24 +1,55 @@
...
+      require "webpacker"
+      ::Webpacker
     end
 
     def self.dev_server_running?
-      return false unless using_webpacker?
+      return false unless using_packer?
 
-      Webpacker.dev_server.running?
+      adapter.dev_server.running?
```
Comment (justin808) [inline lib/react_on_rails/packer_utils.rb]: rename to `packer` from `adapter`

File: lib/react_on_rails/packs_generator.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@ module ReactOnRails
   # rubocop:disable Metrics/ClassLength
   class PacksGenerator
     CONTAINS_CLIENT_OR_SERVER_REGEX = /\.(server|client)($|\.)/
-    MINIMUM_SHAKAPACKER_VERSION = [6, 5, 1].freeze
+    MINIMUM_SHAKAPACKER_VERSION = [7, 2, 1].freeze
```
Comment (justin808) [inline lib/react_on_rails/packs_generator.rb]: undo this

File: lib/react_on_rails/configuration.rb
Comment (justin808) [inline lib/react_on_rails/configuration.rb:1]: double check all these changes

File: lib/react_on_rails/configuration.rb [hunk 5]
```diff
@@ -256,27 +258,23 @@ def raise_missing_components_subdirectory
...
     def shakapacker_precompile?
-      return Webpacker.config.webpacker_precompile? if ReactOnRails::WebpackerUtils.using_shakapacker_6?
-
-      Webpacker.config.shakapacker_precompile?
+      Shakapacker.config.shakapacker_precompile?
     end
 
     def shakapacker_clean_task
-      ReactOnRails::WebpackerUtils.using_shakapacker_6? ? "webpacker:clean" : "shakapacker:clean"
+      "shakapacker:clean"
```
Comment (justin808) [inline lib/react_on_rails/configuration.rb]: why?

File: lib/react_on_rails/configuration.rb [hunk 6]
```diff
@@ -256,27 +258,23 @@ def raise_missing_components_subdirectory
     end
 
     def shakapacker_precompile?
-      return Webpacker.config.webpacker_precompile? if ReactOnRails::WebpackerUtils.using_shakapacker_6?
-
-      Webpacker.config.shakapacker_precompile?
+      Shakapacker.config.shakapacker_precompile?
```
Comment (justin808) [inline lib/react_on_rails/configuration.rb]: why?

File: CHANGELOG.md [hunk 7]
```diff
@@ -18,10 +18,13 @@
 ### [Unreleased]
 Changes since the last non-beta release.
 
+#### Fixed
+- Resolves possible thrown error for precompile checks with Shakapacker. [PR 1622](https://github.com/shakacode/react_on_rails/pull/1622) by [adriangohjw](https://github.com/adriangohjw)
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: Enables React on Rails to be used with any version of Shakapacker and maybe most versions of Webpacker.

[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md]: and by judahmeek

**Review by justin808 — CHANGES_REQUESTED** [review]

**Comment by justin808** [comment]
@G-Rath any opinions on this PR?

**Comment by vaukalak** [comment]
Closing, since the issue has been fixed by: https://github.com/shakacode/react_on_rails/pull/1629
CC: @justin808

**Comment by Judahmeek** [comment]
@vaukalak #1629 doesn't resolve all the incompatibilities regarding Shakapacker v8.

It unfortunately doesn't even resolve the generator issues because the Examples job was failing to actually run the expected test suite due to an id mismatch between [changed-files-specific](https://github.com/shakacode/react_on_rails/blob/master/.github/workflows/examples.yml#L17) and [steps.changed-files.outputs.any_changed](https://github.com/shakacode/react_on_rails/blob/master/.github/workflows/examples.yml#L80)

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -55,36 +54,39 @@ jobs:
...
       run: cd spec/dummy && yarn install --no-progress --no-emoji
+    - run: cd spec/dummy && yarn add "shakapacker@${{ matrix.versions == 'oldest' && '6.6.0' || '8.0.0' }}"
     - name: Save dummy app ruby gems to cache
       uses: actions/cache@v3
       with:
         path: spec/dummy/vendor/bundle
-        key: v5-dummy-app-gem-cache-${{ hashFiles('spec/dummy/Gemfile.lock') }}
+        key: dummy-app-gem-cache-${{ hashFiles('react_on_rails.gemspec') }}-${{ hashFiles('Gemfile.development_dependencies') }}-${{ matrix.versions }}
+    - name: Dynamically add the right shakapacker version to the Gemfile
+      run: echo "gem 'shakapacker', '~> ${{ matrix.versions == 'oldest' && '6.6.0' || '8.0.0' }}'" >> spec/dummy/Gemfile
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/main.yml]: Let's have the latest version back in spec/dummy and have the CI script modify the gemfile to test the older version.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -93,12 +95,12 @@ jobs:
...
       uses: ruby/setup-ruby@v1
       with:
-        ruby-version: ${{ matrix.ruby }}
+        ruby-version: ${{ matrix.versions == 'oldest' && '3.0' || '3.3' }}
         bundler: 2.5.9
     - name: Setup Node
       uses: actions/setup-node@v3
       with:
-        node-version: ${{ matrix.node }}
+        node-version: ${{ matrix.versions == 'oldest' && '16' || '20' }}
```
Comment (justin808) [inline .github/workflows/main.yml:102]: 22 not 20

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -117,33 +119,37 @@ jobs:
...
-        yarn run eslint -v
         sudo yarn global add yalc
     - name: yalc publish for react-on-rails
       run: yalc publish
     - name: yalc add react-on-rails
       run: cd spec/dummy && yalc add react-on-rails
+    - run: cd spec/dummy && yarn add "shakapacker@${{ matrix.versions == 'oldest' && '6.6.0' || '8.0.0' }}"
+    - name: Dynamically add the right shakapacker version to Gemfiles
+      run: |
+        echo "gem 'shakapacker', '~> ${{ matrix.versions == 'oldest' && '6.6.0' || '8.0.0' }}'" >> Gemfile
```
Comment (justin808) [inline .github/workflows/main.yml]: Have https://github.com/shakacode/react_on_rails/blob/master/Gemfile.development_dependencies#L3 use the latest, and use a couple lines of bash code to change to the 6.6.0 when needed.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -153,11 +142,17 @@ yarn test
...
 
 ```sh
-cd react_on_rails/spec/dummy
+cd react_on_rails
+npm -g i yalc
+bundle && yarn
+yalc publish
+cd spec/dummy
+bundle && yarn
+yarn add shakapacker
```
Comment (justin808) [inline CONTRIBUTING.md]: remove

Comment (justin808) [reply CONTRIBUTING.md]: add instructions on how to locally test the older shakapacker

File: lib/react_on_rails/packer_utils.rb [hunk 12]
```diff
@@ -1,24 +1,55 @@
 # frozen_string_literal: true
 
 module ReactOnRails
-  module WebpackerUtils
+  module PackerUtils
+    def self.using_packer?
+      using_shakapacker? || using_webpacker?
+    end
+
+    def self.using_shakapacker?
```
Comment (justin808) [inline lib/react_on_rails/packer_utils.rb]: `using_shakapacker_const`

File: lib/react_on_rails/packer_utils.rb [hunk 13]
```diff
@@ -1,24 +1,55 @@
...
       return @using_webpacker if defined?(@using_webpacker)
 
-      @using_webpacker = ReactOnRails::Utils.gem_available?("webpacker") ||
-                         ReactOnRails::Utils.gem_available?("shakapacker")
+      @using_webpacker = (ReactOnRails::Utils.gem_available?("shakapacker") &&
+                          shakapacker_version_as_array[0] <= 6) ||
+                         ReactOnRails::Utils.gem_available?("webpacker")
+    end
+
+    def self.packer_type
```
Comment (justin808) [inline lib/react_on_rails/packer_utils.rb:24]: memoize?

File: lib/react_on_rails/packs_generator.rb [hunk 14]
```diff
@@ -102,7 +102,7 @@ def generated_server_bundle_file_path
 
       generated_interim_server_bundle_path = server_bundle_entrypoint.sub(".js", "-generated.js")
       generated_server_bundle_file_name = component_name(generated_interim_server_bundle_path)
-      source_entrypoint_parent = Pathname(ReactOnRails::WebpackerUtils.webpacker_source_entry_path).parent
+      source_entrypoint_parent = Pathname(ReactOnRails::PackerUtils.packer_source_entry_path).parent
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/react_on_rails/packs_generator.rb:105]: remove

spec/dummy/bin/webpacker
spec/dummy/bin/webpacker-dev-server

and have CI copy these files from shakapacker and remove the bin/shakapacker ones.

File: package.json [hunk 15]
```diff
@@ -70,9 +70,7 @@
     "prerelease": "yarn run check && yarn run clean && yarn run build",
     "release:patch": "node_package/scripts/release patch",
     "release:minor": "node_package/scripts/release minor",
-    "release:major": "node_package/scripts/release major",
-    "dummy:install": "rake dummy_apps",
-    "dummy:spec": "rake run_rspec:dummy"
```
Comment (justin808) [inline package.json:75]: revert

File: rakelib/dummy_apps.rake [hunk 16]
```diff
@@ -1,30 +0,0 @@
...
-    sh_in_dir(dummy_app_dir, "bundle exec rake react_on_rails:generate_packs")
-  end
-
-  task dummy_apps: %i[dummy_app node_package generate_packs] do
-    puts "Prepared all Dummy Apps"
-  end
-end
-
-desc "Prepares all dummy apps by installing dependencies"
-task dummy_apps: ["dummy_apps:dummy_apps"]
```
Comment (justin808) [inline rakelib/dummy_apps.rake]: revert

File: rakelib/run_rspec.rake [hunk 17]
```diff
@@ -19,28 +26,43 @@ namespace :run_rspec do
   end
 
   desc "Runs dummy rspec with turbolinks"
-  task dummy: ["dummy_apps:dummy_app"] do
+  task :dummy do
```
Comment (justin808) [inline rakelib/run_rspec.rake]: revert

File: spec/dummy/bin/webpacker
Comment (justin808) [inline spec/dummy/bin/webpacker:1]: can get added for CI by copying file from shakapacker v6 gem

File: spec/dummy/bin/shakapacker-dev-server [hunk 18]
```diff
@@ -8,7 +8,7 @@ ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../../Gemfile",
   Pathname.new(__FILE__).realpath)
 
 require "bundler/setup"
-
+require "rake"
```
Comment (justin808) [inline spec/dummy/bin/shakapacker-dev-server:11]: why? or revert?

File: spec/dummy/config/webpack/commonWebpackConfig.js [hunk 19]
```diff
@@ -1,11 +1,14 @@
 // Common configuration applying to client and server configuration
-const { generateWebpackConfig, merge } = require('shakapacker');
+const { generateWebpackConfig, webpackConfig: v6WebpackConfig, merge } = require('shakapacker');
+
+const generateWebpackConfigAlias = generateWebpackConfig ? generateWebpackConfig : () => undefined;
+
+const baseClientWebpackConfig = v6WebpackConfig ? v6WebpackConfig : generateWebpackConfigAlias();
```
Comment (justin808) [inline spec/dummy/config/webpack/commonWebpackConfig.js]: Revert any complexities here and have CI change the file if needed for tests on 6.6 shakapacker

File: spec/dummy/config/webpacker.yml [hunk 20]
```diff
@@ -0,0 +1,73 @@
+# Note: You must restart bin/shakapacker-dev-server for changes to take effect
```
Comment (justin808) [inline spec/dummy/config/webpacker.yml]: revert this file

File: spec/dummy/spec/helpers/react_on_rails_helper_spec.rb [hunk 21]
```diff
@@ -3,13 +3,15 @@
 require "rails_helper"
 require "support/script_tag_utils"
 
+Packer = Object.const_get(ReactOnRails::PackerUtils.packer_type.capitalize)
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/dummy/spec/helpers/react_on_rails_helper_spec.rb:6]: Can we try calling const_get in PackerUtils?

File: spec/dummy/spec/rake/assets_precompile_rake_spec.rb [hunk 22]
```diff
@@ -3,23 +3,28 @@
 require "rake"
 
 require "rails_helper"
+require "react_on_rails"
 
 describe "rake assets:precompile task" do
   it "doesn't show deprecation message for using webpacker:clean task" do
     allow(ENV).to receive(:[]).with(anything).and_call_original
     allow(ENV).to receive(:[]).with("SHAKAPACKER_PRECOMPILE").and_return("false")
+    allow(ENV).to receive(:[]).with("WEBPACKER_PRECOMPILE").and_return("false")
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/dummy/spec/rake/assets_precompile_rake_spec.rb:12]: Yes, when using option 1 to stub environment variables in RSpec tests, it's a good practice to revert the original value after the test is completed. This ensures that the stubbed value doesn't affect other tests or the overall test environment.

You can use RSpec's `before` and `after` hooks to set up and clean up the stubbed environment variable. Here's an example:

```ruby
describe "My test" do
  before do
    @original_value = ENV["VARIABLE_NAME"]
    allow(ENV).to receive(:[]).with("VARIABLE_NAME").and_return("stubbed_value")
  end

  after do
    allow(ENV).to receive(:[]).with("VARIABLE_NAME").and_call_original
    ENV["VARIABLE_NAME"] = @original_value
  end

  it "tests something" do
    # Your test code here
  end
end
```

In this example:

1. The `before` hook is used to store the original value of the environment variable in an instance variable (`@original_value`) before stubbing it. This allows you to restore the original value later.

2. The `allow` method is used to stub the `ENV` object to return the desired stubbed value when accessed with the specific variable name.

3. The `after` hook is used to clean up the stubbed environment variable. It first resets the stubbed behavior by calling `allow(ENV).to receive(:[]).with("VARIABLE_NAME").and_call_original`, which tells RSpec to revert to the original behavior of `ENV` for that variable.

4. Finally, the original value stored in `@original_value` is assigned back to the environment variable using `ENV["VARIABLE_NAME"] = @original_value`.

By following this approach, you ensure that the stubbed environment variable is properly cleaned up and reverted to its original value after each test, preventing any unintended side effects on other tests.

File: spec/dummy/spec/rake/assets_precompile_rake_spec.rb
Comment (justin808) [inline spec/dummy/spec/rake/assets_precompile_rake_spec.rb:1]: Overall, add a rake task to modify the files of spec/dummy to test Shakapacker 6.6.0. Check before running the task that there are no uncommitted files and error if there are.

**Review by justin808 — CHANGES_REQUESTED** [review]

**Review by justin808 — APPROVED** [review]
LGTM!

**Pull request merged**